can/zynqmp-can/zynqmp-can.c

```diff
@@ -0,0 +1,552 @@
+		debug("zynqmp-can: failed to configure desired baud rate\n");
+		return EXIT_FAILURE;
+	}
+
+	/* Initialize the clock */
+	if (init_clock(can_id) < 0) {
+		debug("zynqmp-can: cannot initialize clocks\n");
+		return EXIT_FAILURE;
+	}
+
+	/* Configure MIO pins */
+	if (config_mio_pin(info[can_id].rxPin) < 0 || config_mio_pin(info[can_id].txPin) < 0) {
+		debug("zynqmp-can: failed to initialize MIO pins\n");
+		return -EINVAL;
+	}
+
+	/* Map access to the peripheral */
+	can.base = mmap(NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE | MAP_PHYSMEM | MAP_ANONYMOUS, -1, info[can_id].base);
+	if (can.base == MAP_FAILED) {
+		debug("zynqmp-can: failed to map peripheral\n");
+		return -ENOMEM;
+	}
+
+	/* Initialize synchronization primitives */
+	if (condCreate(&can.cond) != EOK) {
+		munmap((void *)can.base, _PAGE_SIZE);
+		debug("zynqmp-can: failed to create conditional variable\n");
+		return -ENOENT;
+	}
+	if (mutexCreate(&can.lock) != EOK) {
+		debug("zynqmp-can: failed to create mutex\n");
+		munmap((void *)can.base, _PAGE_SIZE);
+		resourceDestroy(can.cond);
+		return -ENOENT;
+	}
+
+	/* Reset the CAN peripheral */
+	if (reset_can_peripheral(can_id) < 0) {
+		debug("zynqmp-can: failed to reset CAN peripheral\n");
+		return -EINVAL;
+	}
+
+	/* Wait for configuration mode */
+	do {
+		reg_val = can.base->SR & (1 << 0);
+	} while (reg_val != 1);
+
+	/* Program the bit sampling clock */
+	can.base->BRPR = timeCfg[timeCfgN].brpr;
+	can.base->BTR = timeCfg[timeCfgN].btr;
+
+	/* Configure the watermark for TX FIFO */
+	can.base->WIR = (CAN_TX_FIFO_WATERMARK << 8);
+
+	/* Register the interrupt handler */
+	interrupt(info[can_id].irq, can_irq_hdl, &can, can.cond, &can.inth);
+
+	/* Enable the controller */
+	can.base->SRR = (1 << 1);
+
+	printf("zynqmp-can: Initialized CAN%u with baud rate: %u kbps, sample point: %s%%\n",
+			can_id,
+			timeCfg[timeCfgN].baudrate_kbps,
+			timeCfg[timeCfgN].sample);
+
+	return 0;
+}
+
+/**
+ * Swaps the endianness of a 32-bit value.
+ */
+uint32_t swap_endianness(uint32_t value)
+{
+	return ((value >> 24) & 0x000000FF) |
+			((value >> 8) & 0x0000FF00) |
+			((value << 8) & 0x00FF0000) |
+			((value << 24) & 0xFF000000);
+}
+
+/**
+ * Sends CAN frames to the hardware TX FIFO.
+ * Blocks if the FIFO is full and blocking is enabled.
+ */
+static int send_frames(zynqmp_canFrame *frames, uint32_t count, bool block, uint32_t *sendFrames)
+{
+	*sendFrames = 0;
+
+	/* Flush all frames to the hardware TX FIFO */
+	for (uint32_t i = 0; i < count; i++) {
+		/* Check if TX FIFO is not full */
+		if (0U != (can.base->SR & (1 << 10))) {
+			if (block) {
+				/* Disable, clear, and enable "Transmit FIFO Watermark Empty" interrupt */
+				can.base->IER &= ~(1 << 13);
+				can.base->ICR = (1 << 13);
+				can.base->IER |= (1 << 13);
+
+				/* Wait for "Transmit FIFO Watermark Empty" interrupt */
+				condWait(can.cond, can.lock, 0);
+			}
+			else {
+				/* Not enough space in the buffer */
+				return -ENOMEM;
```

Comment:
More suitable for non-block would-block situation.
```suggestion
				return -EAGAIN;
```